src/ORM/Marshaller.php

```diff
@@ -55,28 +55,59 @@ public function __construct(Table $table)
     }
 
     /**
-     * Build the map of property => association names.
+     * Build the map of property => marshalling callable.
      *
+     * @param array $data The data being marshalled.
      * @param array $options List of options containing the 'associated' key.
      * @return array
      */
-    protected function _buildPropertyMap($options)
+    protected function _buildPropertyMap($data, $options)
     {
-        if (empty($options['associated'])) {
-            return [];
+        $map = [];
+        $schema = $this->_table->schema();
+
+        // Is a concrete column?
+        foreach (array_keys($data) as $prop) {
+            $columnType = $schema->columnType($prop);
+            if ($columnType) {
+                $map[$prop] = function ($value, $entity) use ($columnType) {
+                    return Type::build($columnType)->marshal($value);
+                };
+            }
         }
 
-        $include = $options['associated'];
-        $map = [];
-        $include = $this->_normalizeAssociations($include);
+        // Map associations
+        $options += ['associated' => []];
+        $include = $this->_normalizeAssociations($options['associated']);
         foreach ($include as $key => $nested) {
             if (is_int($key) && is_scalar($nested)) {
                 $key = $nested;
                 $nested = [];
             }
             $assoc = $this->_table->association($key);
-            if ($assoc) {
-                $map[$assoc->property()] = ['association' => $assoc] + $nested + ['associated' => []];
+            if (!$assoc) {
+                continue;
+            }
+            if (isset($options['forceNew'])) {
+                $nested['forceNew'] = $options['forceNew'];
+            }
+            $new = function ($value, $entity) use ($assoc, $nested) {
+                $options = $nested + ['associated' => []];
+                return $this->_marshalAssociation($assoc, $value, $options);
+            };
+            $merge = function ($value, $entity) use ($assoc, $nested) {
+                $options = $nested + ['associated' => []];
+                return $this->_mergeAssociation($entity->get($assoc->property()), $assoc, $value, $options);
+            };
```

Comment:
I had hoped to avoid the extra layer of closures, but the different code paths for new/merge required it. I also had to create closures for each property being marshalled, as I didn't want to end up passing extra parameters to Type instance's marshal methods. It is possible that passing the extra argument is less harmful than the additional function calls though.

Are you suggesting we inline `_mergeAssociation` and `_marshalAssociation` here?